File: backend/app/capability_platform/manifest.py
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any


SHARED_DIR = Path(__file__).resolve().parents[3] / "shared"
MANIFEST_PATH = SHARED_DIR / "v3_capabilities.json"
V4_BROWSER_MANIFEST_PATH = SHARED_DIR / "v4_browser_capabilities.json"


@lru_cache(maxsize=1)
def load_capability_manifest() -> list[dict[str, Any]]:
    with MANIFEST_PATH.open("r", encoding="utf-8") as handle:
        data = json.load(handle)
    if not isinstance(data, list):
        raise ValueError("capability manifest must be a list")
    return data


@lru_cache(maxsize=1)
def load_v4_browser_capability_manifest() -> list[dict[str, Any]]:
    with V4_BROWSER_MANIFEST_PATH.open("r", encoding="utf-8") as handle:
        data = json.load(handle)
    if not isinstance(data, list):
        raise ValueError("v4 browser capability manifest must be a list")
    required = {
        "capability_id",
        "version",
        "category",
        "description",
        "dependencies",
        "feature_flag",
        "maturity_level",
        "target_maturity_level",
        "supported_browsers",
        "supported_websites",
        "benchmarks",
        "metrics",
        "known_limitations",
        "rollout_status",
        "safety_constraints",
        "failure_classes",
        "owner",
    }
    for index, entry in enumerate(data):
        if not isinstance(entry, dict):
            raise ValueError(f"v4 browser capability at index {index} must be an object")
        missing = required.difference(entry)
        if missing:
            raise ValueError(
                f"v4 browser capability {entry.get('capability_id', index)!r} missing fields: {sorted(missing)}"
            )
    return data

File: extension/src/capabilities/manifest.ts
export type CapabilityHealthStatus = 'available' | 'degraded' | 'unavailable'

export interface CapabilityManifestEntry {
  id: string
  version: string
  purpose: string
  health: CapabilityHealthStatus
  permissions?: string[]
  constraints?: string[]
  environments?: string[]
}

import capabilityManifest from '../../../shared/v3_capabilities.json'
import v4BrowserCapabilityManifest from '../../../shared/v4_browser_capabilities.json'

export interface BrowserCapabilityManifestEntry {
  capability_id: string
  version: string
  category: string
  description: string
  dependencies: string[]
  feature_flag: string
  maturity_level: number
  target_maturity_level: number
  supported_browsers: string[]
  supported_websites: string[]
  site_adapters: string[]
  benchmarks: string[]
  metrics: string[]
  known_limitations: string[]
  rollout_status: string
  safety_constraints: string[]
  failure_classes: string[]
  owner: string
}

export const EXISTING_BROWSER_CAPABILITIES: CapabilityManifestEntry[] = capabilityManifest.map((capability) => ({
  ...capability,
  health: 'available',
}))

export function compactCapabilityManifest(): CapabilityManifestEntry[] {
  return [...EXISTING_BROWSER_CAPABILITIES]
}

export const V4_WAVE1_BROWSER_CAPABILITIES: BrowserCapabilityManifestEntry[] =
  v4BrowserCapabilityManifest as BrowserCapabilityManifestEntry[]

export function compactV4BrowserCapabilityManifest(): Array<{
  capability_id: string
  version: string
  feature_flag: string
  maturity_level: number
  rollout_status: string
}> {
  return V4_WAVE1_BROWSER_CAPABILITIES.map((capability) => ({
    capability_id: capability.capability_id,
    version: capability.version,
    feature_flag: capability.feature_flag,
    maturity_level: capability.maturity_level,
    rollout_status: capability.rollout_status,
  }))
}
